design-system/src/theme-provider.tsx

```diff
@@ -110,17 +112,47 @@ type TUseThemeResult = {
   ) => Old | New;
   /** @deprecated */
   isNewTheme: boolean;
+  isRebrandingTheme: boolean;
 };
-const useTheme = (_parentSelector = defaultParentSelector): TUseThemeResult => {
+const useTheme = (parentSelector = defaultParentSelector): TUseThemeResult => {
+  const [theme, setTheme] = useState<ThemeName>('default');
+  const parentSelectorRef = useRef(parentSelector);
+
+  const mutationChangeCallback = useCallback((mutationList) => {
+    // We expect only a single element in the mutation list as we configured the
+    // observer to only listen to `data-theme` changes.
+    const [mutationEvent] = mutationList;
+    setTheme((mutationEvent.target as HTMLElement).dataset.theme as ThemeName);
+  }, []);
+
+  useMutationObserver(parentSelector(), mutationChangeCallback, {
+    attributes: true,
+    attributeFilter: ['data-theme'],
+  });
+
   const themedValue: TUseThemeResult['themedValue'] = useCallback(
-    (_defaultThemeValue, newThemeValue) => newThemeValue,
-    []
+    (defaultThemeValue, newThemeValue) =>
+      theme === 'default' ? defaultThemeValue : newThemeValue,
+    [theme]
   );
 
+  // If we use 'useLayoutEffect' here, we would be trying to read the
+  // data attribute before it gets set from the effect in the ThemeProvider
+  useEffect(() => {
+    // We need to read the current theme after the provider is rendered
+    // to have the actual selected theme (calculated client-side) in the
+    // hook local state
+    const nextTheme = parentSelectorRef.current()?.dataset.theme as ThemeName;
+    if (nextTheme) {
+      setTheme(nextTheme);
+    }
+  }, []);
+
   return {
     theme: 'default',
     themedValue,
-    isNewTheme: true,
+    isNewTheme: false,
```

Comment:
I'm not sure, can this be reused? Or could/should that be removed?

Reply:
From my point of view, I'm not expecting us to actually need neither `isNewTheme` nor `themedValue` this time.

Perhaps we can double check we're not using them in the MC internal applications (monorepo and satellites) and remove them.
If I remember correctly, removing its usage was part of the MC cleanup work we did after last year's MC redesign.

Reply:
I think `isNewTheme` / `themedValue` will be needed to apply those changes 🤔
<img width="1306" alt="Screenshot 2024-02-09 at 11 01 29" src="https://github.com/commercetools/ui-kit/assets/49066275/f0170875-2afd-4886-86e5-3fe1855ce250">

About `isNewTheme`, yes it's been cleaned up, but in here only deprecated, so I wonder if removing it would be any kind of breaking change.

Reply:
ℹ️ Following the discussion with Carlos, the component-level values should be implemented using designated design tokens. While the `themedValue` helper remains available, its usage is encouraged only where necessary for applying new colors in satellite repositories.